### PR TITLE
View model inject

### PR DESCRIPTION
**Pull request by shanuka**
View model inject

### CODE REVIEW

**Comment by shanuka** [comment]
View model inject

File: GithubBrowserSample/app/src/main/java/com/android/example/github/ui/repo/RepoFragment.kt [hunk 1]
```diff
@@ -49,6 +49,7 @@ class RepoFragment : Fragment(), Injectable {
     @Inject
     lateinit var viewModelFactory: ViewModelProvider.Factory
 
+    @Inject
```
Comment (dazza5000) [inline GithubBrowserSample/app/src/main/java/com/android/example/github/ui/repo/RepoFragment.kt]: I think that if we inject the ViewModels in this manner, then we will lose the retention of ViewModels during configuration changes.

`    /**
     * Returns an existing ViewModel or creates a new one in the scope (usually, a fragment or
     * an activity), associated with this {@code ViewModelProvider}.
     * <p>
     * The created ViewModel is associated with the given scope and will be retained
     * as long as the scope is alive (e.g. if it is an activity, until it is
     * finished or process is killed).
     *
     * @param key        The key to use to identify the ViewModel.
     * @param modelClass The class of the ViewModel to create an instance of it if it is not
     *                   present.
     * @param <T>        The type parameter for the ViewModel.
     * @return A ViewModel that is an instance of the given type {@code T}.
     */
    @NonNull
    @MainThread
    public <T extends ViewModel> T get(@NonNull String key, @NonNull Class<T> modelClass) {
        ViewModel viewModel = mViewModelStore.get(key);

        if (modelClass.isInstance(viewModel)) {
            //noinspection unchecked
            return (T) viewModel;
        } else {
            //noinspection StatementWithEmptyBody
            if (viewModel != null) {
                // TODO: log a warning.
            }
        }

        viewModel = mFactory.create(modelClass);
        mViewModelStore.put(key, viewModel);
        //noinspection unchecked
        return (T) viewModel;
    }`

Comment (shanuka) [reply GithubBrowserSample/app/src/main/java/com/android/example/github/ui/repo/RepoFragment.kt]: Thank you very much for your reply.  why do you create following code? 
`import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.android.example.github.ui.repo.RepoViewModel
import com.android.example.github.ui.search.SearchViewModel
import com.android.example.github.ui.user.UserViewModel
import com.android.example.github.viewmodel.GithubViewModelFactory

import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Suppress("unused")
@Module
abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(UserViewModel::class)
    abstract fun bindUserViewModel(userViewModel: UserViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(SearchViewModel::class)
    abstract fun bindSearchViewModel(searchViewModel: SearchViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(RepoViewModel::class)
    abstract fun bindRepoViewModel(repoViewModel: RepoViewModel): ViewModel

    @Binds
    abstract fun bindViewModelFactory(factory: GithubViewModelFactory): ViewModelProvider.Factory
}`

**Comment by dazza5000** [comment]
I'm thinking that this PR goes against the intent of using ViewModel.Factory

**Comment by dazza5000** [comment]
that code is from the viewmodel source code - it shows that there is caching of viewmodel objects which might be lost if we migrate to injeting viewmodels

**Comment by shanuka** [comment]
im asking what the point of having ViewModelModule?

**Comment by dazza5000** [comment]
To provide the map that is injected into the factory

https://blog.kotlin-academy.com/understanding-dagger-2-multibindings-viewmodel-8418eb372848

**Comment by ianhanniballake** [comment]
You must use `ViewModel.Factory` to get instances of your ViewModels.

**Pull request closed without merging**